File: edge-pi/node/capture.js
/**
 * COAV Edge Capture — Node.js (USB webcam + dump1090 ADS-B → Azure Event Hub)
 *
 * Captures one JPEG frame per second from a USB webcam, reads ADS-B position data
 * from a local dump1090 SBS stream, and sends paired ADSB_TELEMETRY + EDGE_VISION_AI
 * events to Azure Event Hub.
 *
 * Contrail inference is delegated to the Python sidecar (inference.py) via
 * child_process.execFile — keeps the hot path in Node while the ML model lives
 * in Python/ONNX Runtime.
 *
 * Install:  npm install
 * Run:      CONN_STR="<eh_conn_str>" node capture.js
 */

'use strict'

const { EventHubProducerClient } = require('@azure/event-hubs')
const { execFile }               = require('child_process')
const net                        = require('net')
const { createCanvas }           = require('canvas')   // node-canvas for frame encoding
const v4l2camera                 = require('v4l2camera') // Linux V4L2 USB webcam
const path                       = require('path')

// ── Config ────────────────────────────────────────────────────────────────────
const CONN_STR      = process.env.CONN_STR       || (() => { throw new Error('CONN_STR required') })()
const EVENTHUB_NAME = 'telemetry-adsb-inbound'
const CAMERA_ID     = process.env.CAMERA_ID      || 'PI-NODE-CAM-01'
const CAMERA_DEV    = process.env.CAMERA_DEV     || '/dev/video0'
const DUMP1090_HOST = process.env.DUMP1090_HOST  || '127.0.0.1'
const DUMP1090_PORT = parseInt(process.env.DUMP1090_PORT || '30003', 10)
const CAPTURE_FPS   = 1                           // 1 frame/sec

const INFERENCE_PY  = path.join(__dirname, '..', 'python', 'inference.py')
const PYTHON_BIN    = process.env.PYTHON_BIN     || 'python3'

// ── ADS-B state ───────────────────────────────────────────────────────────────
/** @type {Map<string, {callsign:string, lat:number, lon:number, alt:number, spd:number, hdg:number, ts:number}>} */
const tracks = new Map()
const TRACK_TTL_MS = 60_000

function parseSbsLine(line) {
  const parts = line.split(',')
  if (parts.length < 22 || parts[0] !== 'MSG') return
  const icao    = parts[4].trim().toUpperCase()
  const msgType = parts[1].trim()
  const now     = Date.now()

  let track = tracks.get(icao) || { callsign: null, lat: null, lon: null,
                                      alt: null, spd: 0, hdg: null, ts: now }
  track.ts = now

  if (msgType === '1' && parts[10].trim())       track.callsign = parts[10].trim()
  if (msgType === '3') {
    if (parts[11]) track.alt = parseInt(parts[11], 10)
    if (parts[14]) track.lat = parseFloat(parts[14])
    if (parts[15]) track.lon = parseFloat(parts[15])
  }
  if (msgType === '4') {
    if (parts[12]) track.spd = Math.round(parseFloat(parts[12]))
    if (parts[13]) track.hdg = parseFloat(parts[13])
  }
  tracks.set(icao, track)
}

function getActiveFlights() {
  const now = Date.now()
  for (const [icao, t] of tracks) {
    if (now - t.ts > TRACK_TTL_MS) tracks.delete(icao)
  }
  return [...tracks.values()].filter(
    t => t.callsign && t.lat !== null && t.lon !== null && t.alt !== null
  )
}

function connectDump1090() {
  const client = new net.Socket()
  let buf = ''

  const reconnect = () => {
    console.warn('[ADS-B] dump1090 disconnected — retrying in 5 s')
    setTimeout(connectDump1090, 5_000)
  }

  client.connect(DUMP1090_PORT, DUMP1090_HOST, () => {
    console.info(`[ADS-B] Connected to dump1090 at ${DUMP1090_HOST}:${DUMP1090_PORT}`)
  })
  client.on('data', chunk => {
    buf += chunk.toString('ascii')
    const lines = buf.split('\n')
    buf = lines.pop()
    lines.forEach(parseSbsLine)
  })
  client.on('close', reconnect)
  client.on('error', () => client.destroy())
}

// ── Camera (V4L2 USB webcam) ──────────────────────────────────────────────────
let cam = null

function initCamera() {
  try {
    cam = new v4l2camera.Camera(CAMERA_DEV)
    cam.configSet({ width: 1280, height: 720 })
    cam.start()
    console.info(`[Camera] V4L2 webcam on ${CAMERA_DEV}`)
  } catch (err) {
    console.warn(`[Camera] V4L2 unavailable (${err.message}) — using synthetic frame`)
    cam = null
  }
}

function captureFrame() {
  return new Promise((resolve) => {
    if (!cam) {
      // Synthetic frame for testing without hardware
      const canvas = createCanvas(1280, 720)
      const ctx    = canvas.getContext('2d')
      ctx.fillStyle = '#87CEEB'
      ctx.fillRect(0, 0, 1280, 720)
      // Draw synthetic contrail
      ctx.strokeStyle = 'rgba(255,255,255,0.8)'
      ctx.lineWidth   = 4
      ctx.beginPath(); ctx.moveTo(100, 200); ctx.lineTo(1180, 180); ctx.stroke()
      resolve(canvas.toBuffer('image/jpeg'))
      return
    }
    cam.capture(frame => {
      // frame is raw YUV — encode to JPEG via canvas
      const canvas = createCanvas(cam.width, cam.height)
      const ctx    = canvas.getContext('2d')
      const img    = ctx.createImageData(cam.width, cam.height)
      cam.toRGB(img.data)
      ctx.putImageData(img, 0, 0)
      resolve(canvas.toBuffer('image/jpeg'))
    })
  })
}

// ── Python inference sidecar ──────────────────────────────────────────────────
const tmpFramePath = '/tmp/coav_edge_frame.jpg'
const { writeFileSync } = require('fs')

/**
 * Writes the JPEG to /tmp, calls `python3 inference.py /tmp/frame.jpg`,
 * parses stdout for contrail_detected and confidence.
 * @returns {Promise<{contrail_detected: boolean, confidence: number}>}
 */
function runInference(jpegBuffer) {
  return new Promise(resolve => {
    try {
      writeFileSync(tmpFramePath, jpegBuffer)
    } catch {
      return resolve({ contrail_detected: false, confidence: 0.0 })
    }
    execFile(PYTHON_BIN, [INFERENCE_PY, tmpFramePath], { timeout: 5_000 }, (err, stdout) => {
      if (err) {
        console.warn('[Inference] Python call failed:', err.message)
        return resolve({ contrail_detected: false, confidence: 0.0 })
      }
      // Parse key: value lines from inference.py stdout
      const detected   = /contrail detected\s*:\s*true/i.test(stdout)
      const confMatch  = stdout.match(/confidence\s*:\s*([\d.]+)/i)
      const confidence = confMatch ? parseFloat(confMatch[1]) : 0.0
      resolve({ contrail_detected: detected, confidence })
    })
  })
}

// ── Validation (mirrors Python TelemetryEvent schema) ─────────────────────────
function isValidCallsign(cs) {
  return typeof cs === 'string' && cs.length >= 3 && cs.length <= 12 && /^[A-Z0-9-]+$/.test(cs)
}

// ── Main loop ──────────────────────────────────────────────────────────────────
async function main() {
  connectDump1090()
  initCamera()

  const producer = new EventHubProducerClient(CONN_STR, EVENTHUB_NAME)
  console.info(`[COAV Edge] Sending to Event Hub: ${EVENTHUB_NAME}`)

  const tick = async () => {
    const jpegBuf  = await captureFrame()
    const infer    = await runInference(jpegBuf)
    const flights  = getActiveFlights()
    const iso      = new Date().toISOString()

    const events = []
    for (const t of flights) {
      if (!isValidCallsign(t.callsign)) continue  // OWASP A03

      const base = {
        flight_id:   t.callsign,
        timestamp:   iso,
        latitude:    +t.lat.toFixed(5),
        longitude:   +t.lon.toFixed(5),
        altitude_ft: t.alt,
        speed_knots: t.spd,
        heading:     t.hdg !== null ? +t.hdg.toFixed(1) : null,
      }
      events.push({ message_type: 'ADSB_TELEMETRY', ...base })
      events.push({
        message_type:      'EDGE_VISION_AI',
        camera_id:         CAMERA_ID,
        contrail_detected: infer.contrail_detected,
        confidence_score:  +infer.confidence.toFixed(3),
        ...base,
      })
    }

    if (events.length > 0) {
      const batch = await producer.createBatch()
      for (const e of events) {
        batch.tryAdd({ body: JSON.stringify(e) })
      }
      await producer.sendBatch(batch)
    }

    console.info(
      `[tick] contrail=${infer.contrail_detected} conf=${infer.confidence.toFixed(2)} ` +
      `flights=${flights.length} events=${events.length}`
    )
  }

  // Run immediately, then every 1/CAPTURE_FPS seconds
  await tick()
  setInterval(tick, Math.round(1000 / CAPTURE_FPS))
}

main().catch(err => { console.error(err); process.exit(1) })

File: edge-pi/python/capture.py
"""
COAV Edge Capture — Raspberry Pi / ACI
Captures frames at 1 FPS, runs contrail inference, reads ADS-B from dump1090,
then sends paired telemetry events to Azure Event Hub.

Event types sent (mirrors emulator.py schema):
  ADSB_TELEMETRY  — position + flight level from dump1090 SBS feed
  EDGE_VISION_AI  — contrail detection result from inference.py

Tested on:
  Raspberry Pi 4 + Camera Module 3 (picamera2)
  Raspberry Pi 4 + USB webcam (OpenCV VideoCapture)

Install: pip install -r requirements.txt
Run:     CONN_STR="<event_hub_conn_str>" python capture.py
"""

import datetime
import json
import logging
import os
import socket
import sys
import threading
import time
from typing import Optional

import cv2
import numpy as np
from azure.eventhub import EventData, EventHubProducerClient
from pydantic import BaseModel, Field, ValidationError

from inference import ContrailDetector

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
logger = logging.getLogger(__name__)

# ── Config ─────────────────────────────────────────────────────────────────────
EVENTHUB_NAME  = "telemetry-adsb-inbound"
CAMERA_ID      = os.environ.get("CAMERA_ID", "PI-CAM-01")
CAPTURE_FPS    = 1                                 # 1 frame/sec — matches emulator cadence
DUMP1090_HOST  = os.environ.get("DUMP1090_HOST", "127.0.0.1")
DUMP1090_PORT  = int(os.environ.get("DUMP1090_PORT", "30003"))   # SBS BaseStation port


# ── Telemetry schema (OWASP A03:2021 — validates before sending) ───────────────
class TelemetryEvent(BaseModel):
    message_type: str
    camera_id: Optional[str]   = None
    flight_id: str             = Field(..., min_length=3, max_length=12,
                                       pattern=r"^[A-Z0-9\-]+$")
    timestamp: str
    latitude: float            = Field(..., ge=-90.0, le=90.0)
    longitude: float           = Field(..., ge=-180.0, le=180.0)
    altitude_ft: int           = Field(..., ge=0, le=60000)
    speed_knots: int           = Field(..., ge=0, le=1000)
    heading: Optional[float]   = Field(None, ge=0.0, le=360.0)
    contrail_detected: Optional[bool]  = None
    confidence_score: Optional[float]  = Field(None, ge=0.0, le=1.0)


# ── ADS-B reader (dump1090 SBS BaseStation format) ────────────────────────────

class AdsbTrack:
    """Accumulates SBS messages for one ICAO hex code into a usable Flight record."""

    def __init__(self, icao: str):
        self.icao       = icao
        self.callsign   : Optional[str]   = None
        self.lat        : Optional[float] = None
        self.lon        : Optional[float] = None
        self.altitude   : Optional[int]   = None
        self.speed      : Optional[int]   = None
        self.heading    : Optional[float] = None
        self.updated_at : float           = time.monotonic()

    def is_complete(self) -> bool:
        return all([self.callsign, self.lat is not None,
                    self.lon is not None, self.altitude is not None])

    def stale(self, max_age_s: float = 60.0) -> bool:
        return (time.monotonic() - self.updated_at) > max_age_s


class Dump1090Reader(threading.Thread):
    """
    Reads the SBS (BaseStation) TCP stream from dump1090 on port 30003.
    Each MSG type 1/3/4 updates the shared track dictionary.
    """

    def __init__(self):
        super().__init__(daemon=True)
        self._tracks: dict[str, AdsbTrack] = {}
        self._lock = threading.Lock()

    def run(self) -> None:
        while True:
            try:
                self._connect_and_read()
            except Exception as exc:
                logger.warning("dump1090 disconnected: %s — retrying in 5 s", exc)
                time.sleep(5)

    def _connect_and_read(self) -> None:
        with socket.create_connection((DUMP1090_HOST, DUMP1090_PORT), timeout=10) as sock:
            logger.info("Connected to dump1090 at %s:%d", DUMP1090_HOST, DUMP1090_PORT)
            buf = ""
            while True:
                chunk = sock.recv(4096).decode("ascii", errors="ignore")
                if not chunk:
                    break
                buf += chunk
                *lines, buf = buf.split("\n")
                for line in lines:
                    self._parse_sbs(line.strip())

    def _parse_sbs(self, line: str) -> None:
        # SBS format: MSG,<msgType>,<sessionID>,<aircraftID>,<hexIdent>,...
        parts = line.split(",")
        if len(parts) < 22 or parts[0] != "MSG":
            return
        icao     = parts[4].strip().upper()
        msg_type = parts[1].strip()

        with self._lock:
            track = self._tracks.setdefault(icao, AdsbTrack(icao))
            track.updated_at = time.monotonic()

            if msg_type == "1":                    # callsign identification
                cs = parts[10].strip()
                if cs:
                    track.callsign = cs
            elif msg_type == "3":                  # airborne position
                try:
                    track.altitude = int(parts[11]) if parts[11] else track.altitude
                    track.lat      = float(parts[14]) if parts[14] else track.lat
                    track.lon      = float(parts[15]) if parts[15] else track.lon
                except ValueError:
                    pass
            elif msg_type == "4":                  # airborne velocity
                try:
                    track.speed   = int(float(parts[12])) if parts[12] else track.speed
                    track.heading = float(parts[13])       if parts[13] else track.heading
                except ValueError:
                    pass

    def get_active_flights(self) -> list[AdsbTrack]:
        with self._lock:
            stale = [k for k, v in self._tracks.items() if v.stale()]
            for k in stale:
                del self._tracks[k]
            return [t for t in self._tracks.values() if t.is_complete()]


# ── Camera capture ─────────────────────────────────────────────────────────────

def open_camera() -> cv2.VideoCapture:
    """Opens picamera2 (Pi Camera Module 3) or first available USB webcam."""
    try:
        from picamera2 import Picamera2          # type: ignore
        picam = Picamera2()
        picam.configure(picam.create_preview_configuration(
            main={"format": "BGR888", "size": (1920, 1080)}
        ))
        picam.start()
        logger.info("Camera: picamera2 (Pi Camera Module 3)")
        return picam                             # duck-typed: has .capture_array()
    except (ImportError, RuntimeError):
        pass

    cap = cv2.VideoCapture(0)
    if not cap.isOpened():
        raise RuntimeError("No camera found — check USB webcam or Pi Camera connection")
    cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1920)
    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 1080)
    logger.info("Camera: USB webcam via OpenCV")
    return cap


def capture_frame(camera) -> Optional[np.ndarray]:
    """Returns BGR frame from either picamera2 or cv2.VideoCapture."""
    try:
        if hasattr(camera, "capture_array"):     # picamera2
            return camera.capture_array("main")  # already BGR
        ret, frame = camera.read()
        return frame if ret else None
    except Exception as exc:
        logger.error("Frame capture failed: %s", exc)
        return None


# ── Main loop ──────────────────────────────────────────────────────────────────

def main() -> None:
    conn_str = os.environ.get("CONN_STR")
    if not conn_str:
        raise RuntimeError("CONN_STR environment variable is required")

    detector = ContrailDetector()
    adsb     = Dump1090Reader()
    adsb.start()

    camera = open_camera()
    producer = EventHubProducerClient.from_connection_string(
        conn_str=conn_str, eventhub_name=EVENTHUB_NAME
    )

    logger.info("COAV Edge Capture started (camera=%s, 1 FPS)", CAMERA_ID)

    try:
        with producer:
            while True:
                loop_start = time.monotonic()

                frame = capture_frame(camera)
                if frame is None:
                    logger.warning("Skipping empty frame")
                    time.sleep(1)
                    continue

                # Run contrail inference
                detection = detector.detect(frame)
                iso = datetime.datetime.now(datetime.timezone.utc).isoformat()

                # Build one event pair per active ADS-B track
                flights = adsb.get_active_flights()
                if not flights:
                    logger.debug("No ADS-B tracks — waiting for dump1090 data")

                events: list[dict] = []
                for track in flights:
                    base = dict(
                        flight_id   = track.callsign,
                        timestamp   = iso,
                        latitude    = round(track.lat, 5),
                        longitude   = round(track.lon, 5),
                        altitude_ft = track.altitude,
                        speed_knots = track.speed or 0,
                        heading     = round(track.heading, 1) if track.heading else None,
                    )
                    events.append(dict(message_type="ADSB_TELEMETRY", **base))
                    events.append(dict(
                        message_type      = "EDGE_VISION_AI",
                        camera_id         = CAMERA_ID,
                        contrail_detected = detection.contrail_detected,
                        confidence_score  = detection.confidence,
                        **base,
                    ))

                # Validate + send
                batch = producer.create_batch()
                for raw in events:
                    try:
                        validated = TelemetryEvent(**raw)
                        batch.add(EventData(validated.model_dump_json()))
                    except ValidationError as ve:
                        logger.error("Validation error (OWASP A03): %s", ve)

                if len(batch) > 0:
                    producer.send_batch(batch)

                logger.info(
                    "tick: contrail=%s conf=%.2f flights=%d events=%d backend=%s",
                    detection.contrail_detected, detection.confidence,
                    len(flights), len(events), detection.backend
                )

                # Maintain 1 FPS
                elapsed = time.monotonic() - loop_start
                time.sleep(max(0.0, 1.0 / CAPTURE_FPS - elapsed))

    except KeyboardInterrupt:
        logger.info("Stopped by user")
    finally:
        if hasattr(camera, "stop"):
            camera.stop()
        elif hasattr(camera, "release"):
            camera.release()


if __name__ == "__main__":
    main()
